runtime-modules/project-token/src/lib.rs

```diff
@@ -0,0 +1,611 @@
+
+        // == MUTATION SAFE ==
+
+        Self::do_deissue_token(token_id);
+        Ok(())
+    }
+
+    /// Retrieve usable (free) balance for token and account
+    /// Preconditions:
+    /// - `token_id` must be a valid token identifier
+    /// - `who` account id must exist for token
+    fn balance(token_id: T::TokenId, who: T::AccountId) -> Result<T::Balance, DispatchError> {
+        let account_info = Self::ensure_account_data_exists(token_id, &who)?;
+
+        Ok(account_info.free_balance)
+    }
+
+    /// Retrieve total current issuance for token
+    /// Preconditions
+    /// - `token_id` must be valid
+    fn current_issuance(token_id: Self::TokenId) -> Result<Self::Balance, DispatchError> {
+        Self::ensure_token_exists(token_id).map(|token_data| token_data.current_total_issuance)
+    }
+}
+
+/// ReservableMultiCurrency trait implementation for Module
+impl<T: Trait> ReservableMultiCurrency<T::AccountId> for Module<T> {
+    type TokenId = T::TokenId;
+    type Balance = T::Balance;
+
+    /// Reserve `amount` of token for `who`
+    /// Preconditions:
+    /// - `token_id` must id
+    /// - `who` must identify valid account for `token_id`
+    /// - `who` free balance must be greater than `amount`
+    ///
+    /// Postconditions:
+    /// - `who` free balance decreased by `amount`
+    /// - `who` reserved balance increased by `amount`
+    /// if `amount` is zero it is equivalent to a no-op
+    fn reserve(token_id: T::TokenId, who: T::AccountId, amount: T::Balance) -> DispatchResult {
+        if amount.is_zero() {
+            return Ok(());
+        }
+
+        // ensure token validity
+        let _ = Self::ensure_token_exists(token_id)?;
+
+        // ensure src account id validity
+        let account_info = Self::ensure_account_data_exists(token_id, &who)?;
+
+        // ensure can freeze amount
+        ensure!(
+            account_info.free_balance >= amount,
+            Error::<T>::InsufficientFreeBalanceForReserving,
+        );
+
+        // == MUTATION SAFE ==
+
+        AccountInfoByTokenAndAccount::<T>::mutate(token_id, &who, |account_data| {
+            account_data.free_balance = account_data.free_balance.saturating_sub(amount);
+            account_data.reserved_balance = account_data.reserved_balance.saturating_add(amount);
+        });
+
+        Self::deposit_event(RawEvent::TokenAmountReservedFrom(token_id, who, amount));
+        Ok(())
+    }
+
+    /// Unreserve `amount` of token for `who`
+    /// Preconditions:
+    /// - `token_id` must id
+    /// - `who` must identify valid account for `token_id`
+    /// - `who` reserved balance must be greater than `amount`
+    ///
+    /// Postconditions:
+    /// - `who` free balance increased by `amount`
+    /// - `who` reserved balance decreased by `amount`
+    /// if `amount` is zero it is equivalent to a no-op
+    fn unreserve(token_id: T::TokenId, who: T::AccountId, amount: T::Balance) -> DispatchResult {
```

Comment:
Why is this a pallet method? Reservation is done by $CRT holders when they want to participate in a revenue split, hence they would only be doing so directly in this pallet, never from another pallet? hence this should at best be a private method?

Reply:
I moved this method to the module `impl` outside of the main `trait`. It will be used for revenue splits

Reply:
in b2726deded2bf71154e0e8e5e8ad01c4fbaaa257